File: jiho/level1/신규아이디추천.java
//https://school.programmers.co.kr/learn/courses/30/lessons/72410
//신규 아이디 추천

package level1;

public class 신규아이디추천 {
    public static void main(String[] args) {
//        String s = "...!@BaT#*..y.abcdefghijklm";
        String s = "z-+.^.";
        String result = solution(s);
        System.out.println(result);
    }

    public static String solution(String new_id) {
        //1단계
        String s = new_id.toLowerCase();

        //2단계
        String sl = "~!@#$%^&*()=+[{]}:?,<>/";
        for(char c : sl.toCharArray()){
            s = s.replace(Character.toString(c),"");
        }

        //3단계
        String s2 = "", before = "";
        for(int i = 0; i<s.length(); i++){
            String now = Character.toString(s.charAt(i));
            if(now.equals(".")){
                s2 += before.equals(".") ? "" : now;
            } else{
                s2 += now;
            }
            before = now;
        }
        s = s2;

        //4단계
        String[] arr = s.split("");

        arr[0] = arr[0].equals(".") ? "" : arr[0];
        arr[arr.length-1] = arr[arr.length-1].equals(".") ? "" : arr[arr.length-1];

        s = String.join("",arr);

        //5단계
        if(s.equals("") || s == null){
            s = "a";
        }

        //6단계
        if(s.length() > 15){
            s = s.substring(0,15);
        }
        arr = s.split("");
        arr[arr.length-1] = arr[arr.length-1].equals(".") ? "" : arr[arr.length-1];
        s = String.join("", arr);

        //7단계
        if(s.length() < 3){
            while(s.length() < 3){
                s += s.charAt(s.length()-1);
            }
        }

        return s;
    }
}

File: ningpop/level1/신규_아이디_추천.py
# 2022.08.10
# 풀이 시간: 29분 36초
# 채점 결과: 정답
# 시간복잡도: O(N)
# 문제 링크: https://school.programmers.co.kr/learn/courses/30/lessons/72410

import re

def solution(new_id: str) -> str:
    modified_new_id = new_id
    
    modified_new_id = modified_new_id.lower()

    modified_new_id = re.sub('[^a-z0-9-_.]', '', modified_new_id)

    modified_new_id = re.sub('[.]+', '.', modified_new_id)

    modified_new_id = re.sub('^[.]|[.]$', '', modified_new_id)

    if modified_new_id == '':
        modified_new_id = 'a'

    if len(modified_new_id) >= 16:
        modified_new_id = modified_new_id[:15]
        if modified_new_id[-1] == '.':
            modified_new_id = modified_new_id[:-1]
    
    if len(modified_new_id) <= 2:
        last_char = modified_new_id[-1]
        while len(modified_new_id) < 3:
            modified_new_id += last_char

    answer = modified_new_id
    return answer

print(solution("...!@BaT#*..y.abcdefghijklm")) # "bat.y.abcdefghi"
print(solution("z-+.^.")) # "z--"
print(solution("=.=")) # "aaa"
print(solution("123_.def")) # "123_.def"
print(solution("abcdefghijklmn.p")) # "abcdefghijklmn"